ecleankernel/file.py

```diff
@@ -137,51 +149,91 @@ def decompress_raw(self) -> bytes:
+                f'Kernel file {self.path} terminates '
+                f'before end of version string')
+        return ret[0].decode()
 
-    def read_internal_version(self) -> str:
-        """Read version from the kernel file"""
-        f = open(self.path, 'rb')
-        f.seek(0x200)
+    def ver_from_BzImage(self, f: BinaryIO) -> str:
+        f.seek(0x200, 1)
         # short seek would result in eof, so read() will return ''
         buf = f.read(0x10)
         if len(buf) != 0x10:
             raise UnrecognizedKernelError(
                 f'Kernel file {self.path} terminates before bzImage '
                 f'header')
-        if buf[2:6] == b'HdrS':
-            offset = struct.unpack_from('H', buf, 0x0e)[0]
-            f.seek(offset - 0x10, 1)
-            buf = f.read(0x100)  # XXX
-            if not buf:
-                raise UnrecognizedKernelError(
-                    f'Kernel file {self.path} terminates before expected '
-                    f'version string position ({offset + 0x200})')
-            ret = buf.split(b' ', 1)
-        else:
-            # If it's not a bzImage it must be a raw binary,
-            # check if it's compressed first
-            b = self.decompress_raw()
-            # unlike with bzImage, the raw kernel binary has no header
-            # that includes the version, so we parse the version message
-            # that appears on boot
-            ver_start = 'Linux version '
-            pos = b.find(ver_start.encode())
-            if pos == -1:
-                raise UnrecognizedKernelError(
-                    f'Kernel file {self.path} does not appear '
-                    f'to have a version string, '
-                    f'or the compression format was not recognized')
-            pos += len(ver_start)
-            sbuf = b[pos:pos + 0x100]
-            ret = sbuf.split(b' ', 1)
-        if len(ret) == 1:
+        offset = struct.unpack_from('H', buf, 0x0e)[0]
+        f.seek(offset - 0x10, 1)
+        buf = f.read(0x100)  # XXX
+        if not buf:
             raise UnrecognizedKernelError(
-                f'Kernel file {self.path} terminates '
-                f'before end of version string')
+                f'Kernel file {self.path} terminates before expected '
+                f'version string position ({offset + 0x200})')
+        ret = buf.split(b' ', 1)
         return ret[0].decode()
 
+    def read_internal_version(self) -> str:
+        """Read version from the kernel file"""
+        f = open(self.path, 'rb')
+
+        # This format is what is described here:
+        # https://www.kernel.org/doc/html/latest/x86/boot.html#the-real-mode-kernel-header
+        if f.read(2) == b'MZ':
+            f.seek(0x3c)
+            header_pos = struct.unpack('B', f.read(1))[0]
+            # Move to COFF file header
+            f.seek(header_pos+4)
+
+            f.seek(2, 1)
+            numberOfSections = struct.unpack('H', f.read(2))[0]
+            f.seek(-4, 1)
```

Comment:
Why are you seeking back more than you seeked forward? This code is so damn confusing! Can't you just read the whole header into a buffer and go from there? There's `struct.unpack_from()` for that.

Reply:
I've commented extensively and reworked some of the odd seeks. How do you feel about it now?